File: src/main/java/me/lokka30/levelledmobs/misc/RequestedLevel.java
package me.lokka30.levelledmobs.misc;

import me.lokka30.levelledmobs.util.Utils;

/**
 * Used as a placeholder for when a number or a number-range is passed from a user argument
 *
 * @author stumper66
 * @since 3.2.0
 */
public class RequestedLevel {

    public int level;
    public int levelRangeMin;
    public int levelRangeMax;
    public boolean hasLevelRange;
    public boolean hadInvalidArguments;

    public void setMinAllowedLevel(final int level) {
        if (this.hasLevelRange && this.levelRangeMin < level) {
            this.levelRangeMin = level;
        } else if (!this.hasLevelRange && this.level < levelRangeMax) {
            this.level = level;
        }
    }

    public void setMaxAllowedLevel(final int level) {
        if (this.hasLevelRange && this.levelRangeMax > level) {
            this.levelRangeMax = level;
        } else if (!this.hasLevelRange && this.level > level) {
            this.level = level;
        }
    }

    public int getLevelMin() {
        if (this.hasLevelRange) {
            return this.levelRangeMin;
        } else {
            return this.level;
        }
    }

    public int getLevelMax() {
        if (this.hasLevelRange) {
            return this.levelRangeMax;
        } else {
            return this.level;
        }
    }

    @SuppressWarnings("BooleanMethodIsAlwaysInverted")
    public boolean setLevelFromString(final String numberOrNumberRange) {
        if (numberOrNumberRange == null || numberOrNumberRange.isEmpty()) {
            return false;
        }

        if (!numberOrNumberRange.contains("-")) {
            if (!Utils.isInteger(numberOrNumberRange)) {
                return false;
            }

            this.level = Integer.parseInt(numberOrNumberRange);
            this.hasLevelRange = false;
            return true;
        }

        final String[] nums = numberOrNumberRange.split("-");
        if (nums.length != 2) {
            return false;
        }

        if (!Utils.isInteger(nums[0].trim()) || !Utils.isInteger(nums[1].trim())) {
            return false;
        }
        this.levelRangeMin = Integer.parseInt(nums[0].trim());
        this.levelRangeMax = Integer.parseInt(nums[1].trim());
        this.hasLevelRange = true;

        return true;
    }

    public String toString() {
        if (this.hasLevelRange) {
            return String.format("%s-%s", this.levelRangeMin, this.levelRangeMax);
        } else {
            return String.valueOf(this.level);
        }
    }
}

File: levelledmobs-plugin/src/main/kotlin/io/github/arcaneplugins/levelledmobs/misc/RequestedLevel.kt
package io.github.arcaneplugins.levelledmobs.misc

import io.github.arcaneplugins.levelledmobs.util.Utils.isInteger

/**
 * Used as a placeholder for when a number or a number-range is passed from a user argument
 *
 * @author stumper66
 * @since 3.2.0
 */
class RequestedLevel {
    var level = 0
    var levelRangeMin = 0
    var levelRangeMax = 0
    var hasLevelRange = false
    var hadInvalidArguments = false

    fun setMinAllowedLevel(level: Int) {
        if (this.hasLevelRange && this.levelRangeMin < level) {
            this.levelRangeMin = level
        } else if (!this.hasLevelRange && this.level < levelRangeMax) {
            this.level = level
        }
    }

    fun setMaxAllowedLevel(level: Int) {
        if (this.hasLevelRange && this.levelRangeMax > level) {
            this.levelRangeMax = level
        } else if (!this.hasLevelRange && this.level > level) {
            this.level = level
        }
    }

    val levelMin: Int
        get() {
            return if (this.hasLevelRange) {
                levelRangeMin
            } else {
                level
            }
        }

    val levelMax: Int
        get() {
            return if (this.hasLevelRange) {
                levelRangeMax
            } else {
                level
            }
        }

    fun setLevelFromString(numberOrNumberRange: String?): Boolean {
        if (numberOrNumberRange.isNullOrEmpty()) {
            return false
        }

        if (!numberOrNumberRange.contains("-")) {
            if (!isInteger(numberOrNumberRange)) {
                return false
            }

            this.level = numberOrNumberRange.toInt()
            this.hasLevelRange = false
            return true
        }

        val nums = numberOrNumberRange.split("-".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        if (nums.size != 2) {
            return false
        }

        if (!isInteger(nums[0].trim { it <= ' ' }) || !isInteger(
            nums[1].trim { it <= ' ' })
        ) {
            return false
        }
        this.levelRangeMin = nums[0].trim { it <= ' ' }.toInt()
        this.levelRangeMax = nums[1].trim { it <= ' ' }.toInt()
        this.hasLevelRange = true

        return true
    }

    override fun toString(): String {
        return if (this.hasLevelRange) {
            "$levelRangeMin-$levelRangeMax"
        } else {
            level.toString()
        }
    }
}